pkg/storage/snapshot/restore.go

```diff
@@ -562,6 +562,16 @@ func (t *vmRestoreTarget) reconcileSpec() (bool, error) {
 	}
 
 	if !t.doesTargetVMExist() {
+		// Drop ControllerRevision names for any referenced preference or instancetype
+		// before we create a new VM. This will force the VirtualMachine controller to
+		// recreate these for the new VM.
+		if newVM.Spec.Instancetype != nil {
+			newVM.Spec.Instancetype.RevisionName = ""
+		}
+		if newVM.Spec.Preference != nil {
+			newVM.Spec.Preference.RevisionName = ""
+		}
+
```

Comment:
what if the instancetype or preference doesn't exist anymore in the cluster? should that impact the restore?

Were any alternative solutions considered? For example, what about making a copy of the instancetype/preference controller revision when the snapshot is created?

Unless we snapshot the controller revisions as well, we might not be getting the exact same VM after a restore.

Reply:
> Another approach might be to add the snapshot and any cloned VirtualMachines as owners of the ControllerRevision to avoid cloning?

Sounds good to me.
Further to our online conversation, I think that in order for it to work we would have to remove owner references from ControllerRevision once the owners are being deleted.

In order to do so, a controller should:
* Be informer of a resource deletion
* Search for ControllerRevisions that are owned by that resource (possibly using an informer's indexer)
* Delete the owner reference from the ControllerRevision

Once no owners are left for the ControllerRevision, it would be automatically deleted.

Or is there a better (i.e. more automatic) way to do that? I wonder how other controllers are coping with this.

Reply:
> > Another approach might be to add the snapshot and any cloned VirtualMachines as owners of the ControllerRevision to avoid cloning?
> 
> Sounds good to me. Further to our online conversation, I think that in order for it to work we would have to remove owner references from ControllerRevision once the owners are being deleted.
> 
> In order to do so, a controller should:
> 
> * Be informer of a resource deletion
> * Search for ControllerRevisions that are owned by that resource (possibly using an informer's indexer)
> * Delete the owner reference from the ControllerRevision
> 
> Once no owners are left for the ControllerRevision, it would be automatically deleted.
> 
> Or is there a better (i.e. more automatic) way to do that? I wonder how other controllers are coping with this.

This was easy enough to do for `VirtualMachineSnapshots` but will require a new `VirtualMachine` finalizer to allow the controller to actually see the resource during deletion. @iholder-redhat I can continue with this approach if you think it's valid but I'm worried I'm introducing too much complexity here given we can just clone the `ControllerRevision` for the `VirtualMachineSnapshot`. Thoughts?

Reply:
I think that using owner references is the better approach here. It has two advantages; it avoids duplicating data on etcd, and it's easier to debug and reason about (revision X is being used by VMs a, b, c etc).

Saying that, I don't think this is urgent at this point. It might be more significant in situations where many VMs are cloned, and even then it's probably not that big of a deal.

I'd say we can keep it to future PRs if we need to.
Thanks @lyarwood for this discussion!

Reply:
> I think that using owner references is the better approach here. It has two advantages; it avoids duplicating data on etcd, and it's easier to debug and reason about (revision X is being used by VMs a, b, c etc).

Agreed but we already create fresh `ControllerRevisions` for each `VirtualMachine` even if they share the same data as existing `ControllerRevisions` for other `VirtualMachines`.

> Saying that, I don't think this is urgent at this point. It might be more significant in situations where many VMs are cloned, and even then it's probably not that big of a deal.
> 
> I'd say we can keep it to future PRs if we need to. Thanks @lyarwood for this discussion!

ACK, yeah I agree that this is a future improvement, I've write up an [issue](https://github.com/kubevirt/kubevirt/issues/8591) and target it at the [`v1beta1` tracker](https://github.com/kubevirt/kubevirt/issues/8235).

For now I'll switch over to cloning the CRs for each `VirtualMachineSnapshot` and each resulting `VirtualMachine`.

Reply:
The current series should address the original offline clone bug by copying the `ControllerRevisions`, associating them with the `VirtualMachineSnapshot` before copying them again during restore and associating them with the target `VirtualMachine`.

I still need to write tests and look into online snapshots as the `vmSnapshotSource` code is doing weird things with the `VirtualMachineInstanceSpec` there that might be incompatible with instancetypes.

Reply:
> I still need to write tests and look into online snapshots as the [`vmSnapshotSource` code is doing weird things with the `VirtualMachineInstanceSpec`](https://github.com/lyarwood/kubevirt/blob/04fa82725efc0e5de5cc9d977570ed0fd370a52f/pkg/storage/snapshot/source.go#L259-L280) there that might be incompatible with instancetypes.

I've pushed the following commit to prompt some discussion on this:

https://github.com/kubevirt/kubevirt/pull/8466/commits/35f82c3462fc62a7f9336cce5626f6d0afed11bd

This code originally came from https://github.com/kubevirt/kubevirt/pull/6642

/cc @ShellyKa13 

Apologies for dragging you into this PR, I'm trying to get my head around why we are fetching the `VirtualMachineInstance` here? Your PR suggested we wanted to include hot plugged disks in the snapshot but AFAICT you're overwriting the disks in the `VirtualMachineInstance` with the original disks from the `VirtualMachine`?

As I've set out in the commit above when instancetypes are used the `VirtualMachineInstanceSpec` in the `VirtualMachienInstance` isn't compatible with the `VirualMachine` anymore. If possible I'd like to stop copying it across while still persisting any hot plugged disks. Thoughts?